File: cache_benchmark.cpp
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

#include "cacheSystem.h"

namespace benchmark {

const std::size_t kRequestsPerTrace = 20000;
const std::vector<unsigned> kRandomSeeds{1, 2, 3};

const std::vector<cache::cacheLevel> kCacheLevels{cache::L1, cache::L2, cache::L3};
const std::vector<double> kLevelAccessCosts{1.0, 5.0, 20.0};

const double kMemoryAccessCost = 100.0;

const std::size_t kTotalCacheCapacity = 64;
const std::vector<std::vector<std::size_t>> kPageCacheDistribution{
    {64},
    {8, 56},
    {16, 48},
    {32, 32},
    {8, 16, 40}
};


enum RequestPattern {
    SmallWorkingSetCycle,
    CapacityBoundaryCycle,
    SequentialScan,
    HotColdAccess,
    HotPagesWithScan,
    ChangingHotPages,
    RepeatedPageBursts,
    UniformRandomAccess
};

struct RequestPatternDescription {
    RequestPattern patternType;
    std::string patternName;
};

const std::vector<RequestPatternDescription> kRequestPatterns{
    {SmallWorkingSetCycle, "small_cycle"},
    {CapacityBoundaryCycle, "boundary_cycle"},
    {SequentialScan, "scan"},
    {HotColdAccess, "hot_cold"},
    {HotPagesWithScan, "hot_scan"},
    {ChangingHotPages, "phase_change"},
    {RepeatedPageBursts, "bursts"},
    {UniformRandomAccess, "uniform"}
};


int generateRandomNumber(std::mt19937& randomGenerator,
                         int minimumValue,
                         int maximumValue) {

    std::uniform_int_distribution<int> numberDistribution(minimumValue, maximumValue);

    return numberDistribution(randomGenerator);
}

int generateHotColdKey(std::mt19937& randomGenerator, int hotPageOffset) {
    if (generateRandomNumber(randomGenerator, 0, 99) < 90) {
        return hotPageOffset + generateRandomNumber(randomGenerator, 0, 7);
    }

    return generateRandomNumber(randomGenerator, 1024, 2047);
}

int generateRequestKey(RequestPattern patternType,
                       std::size_t requestIndex,
                       int previousRequestKey,
                       std::mt19937& randomGenerator) {

    switch (patternType) {
        case SmallWorkingSetCycle:
            return static_cast<int>(requestIndex % 32);

        case CapacityBoundaryCycle:
            return static_cast<int>(requestIndex % 65);

        case SequentialScan:
            return static_cast<int>(requestIndex % 256);

        case HotColdAccess:
            return generateHotColdKey(randomGenerator, 0);

        case HotPagesWithScan:
            if (requestIndex % 512 < 384) {
                return generateRandomNumber(randomGenerator, 0, 7);
            }

            return 1024 + static_cast<int>(requestIndex);

        case ChangingHotPages:
            return generateHotColdKey(
                randomGenerator, static_cast<int>((requestIndex / 2000) % 4) *
                                 static_cast<int>(kTotalCacheCapacity));

        case RepeatedPageBursts:
            if (requestIndex % 16 == 0) {
                return generateRandomNumber(randomGenerator, 0, 255);
            }

            return previousRequestKey;

        case UniformRandomAccess:
            return generateRandomNumber(randomGenerator, 0, 255);

        default:
            throw std::invalid_argument("Unknown pattern");
    }
}

std::vector<int> generateRequestSequence(RequestPattern patternType, unsigned randomSeed) {
    std::mt19937 randomGenerator(randomSeed);
    std::vector<int> requestSequence;
    requestSequence.reserve(kRequestsPerTrace);

    int previousRequestKey = 0;

    for (std::size_t requestIndex = 0; requestIndex < kRequestsPerTrace; ++requestIndex) {
        const int requestedKey = generateRequestKey(
            patternType, requestIndex, previousRequestKey, randomGenerator);

        requestSequence.push_back(requestedKey);
        previousRequestKey = requestedKey;
    }

    return requestSequence;
}


std::vector<cache::cacheDescription> createLevelVariants(std::size_t levelCapacity,
                                                       cache::cacheLevel cacheLevel) {

    std::vector<cache::cacheDescription> levelVariants(5);

    levelVariants[0].strategy = cache::C_LRU;
    levelVariants[1].strategy = cache::C_LFU;
    levelVariants[2].strategy = cache::C_ARC;
    levelVariants[3].strategy = cache::C_2Q;
    levelVariants[4].strategy = cache::C_LIRS;

    for (auto& levelDescription : levelVariants) {
        levelDescription.size = levelCapacity;
        levelDescription.level = cacheLevel;
    }

    return levelVariants;
}

std::vector<cache::cacheSystemParams> appendCacheLevel(
    const std::vector<cache::cacheSystemParams>& cacheConfigurations,
    std::size_t levelCapacity,
    cache::cacheLevel cacheLevel) {

    std::vector<cache::cacheSystemParams> expandedConfigurations;

    for (const auto& levelDescription : createLevelVariants(levelCapacity, cacheLevel)) {
        for (const auto& configuration : cacheConfigurations) {
            cache::cacheSystemParams extendedConfiguration = configuration;
            extendedConfiguration.levels.push_back(levelDescription);

            expandedConfigurations.push_back(extendedConfiguration);
        }
    }

    return expandedConfigurations;
}

void checkLevelCapacities(const std::vector<std::size_t>& levelCapacities) {
    if (levelCapacities.empty() || levelCapacities.size() > kCacheLevels.size()) {
        throw std::invalid_argument("Use 1 to 3 levels");
    }

    std::size_t totalCapacity = 0;

    for (const auto levelCapacity : levelCapacities) {
        if (levelCapacity < 4) {
            throw std::invalid_argument("Use at least 4 slots per level");
        }

        totalCapacity += levelCapacity;
    }

    if (totalCapacity != kTotalCacheCapacity) {
        throw std::invalid_argument("The total cache capacity must be 64");
    }
}

std::vector<cache::cacheSystemParams> createCacheConfigurations() {
    std::vector<cache::cacheSystemParams> allConfigurations;

    for (const auto& levelCapacities : kPageCacheDistribution) {
        checkLevelCapacities(levelCapacities);

        std::vector<cache::cacheSystemParams> cacheConfigurations(1);

        for (std::size_t levelIndex = 0; levelIndex < levelCapacities.size(); ++levelIndex) {
            cacheConfigurations = appendCacheLevel(
                cacheConfigurations, levelCapacities[levelIndex], kCacheLevels[levelIndex]);
        }

        for (const auto& configuration : cacheConfigurations) {
            allConfigurations.push_back(configuration);
        }
    }

    return allConfigurations;
}

std::string getConfigurationName(const cache::cacheSystemParams& configuration) {
    std::string configurationName;

    for (const auto& levelDescription : configuration.levels) {
        if (!configurationName.empty()) {
            configurationName += "+";
        }

        switch (levelDescription.strategy) {
            case cache::C_LRU:
                configurationName += "LRU";
                break;

            case cache::C_LFU:
                configurationName += "LFU";
                break;

            case cache::C_ARC:
                configurationName += "ARC";
                break;

            case cache::C_2Q:
                configurationName += "2Q";
                break;

            case cache::C_LIRS:
                configurationName += "LIRS";
                break;

            default:
                throw std::invalid_argument("Unknown strategy");
        }

        configurationName += std::to_string(levelDescription.size);
    }

    return configurationName;
}


uint32_t generatePageValue(int pageKey) {
    return static_cast<uint32_t>(pageKey) * 2654435761u;
}

void checkSystemStatistics(const cache::CacheSystemStats& systemStatistics,
                           std::size_t requestCount,
                           std::size_t memoryLoadCount,
                           std::size_t levelCount) {

    if (systemStatistics.levels.size() != levelCount) {
        throw std::runtime_error("Wrong level count");
    }

    std::size_t expectedLevelRequests = requestCount;
    std::size_t totalHitCount = 0;

    for (const auto& level : systemStatistics.levels) {
        const auto& levelStatistics = level.stats;

        if (levelStatistics.amountRequests != expectedLevelRequests ||
            levelStatistics.amountHits + levelStatistics.amountMisses != expectedLevelRequests) {
            throw std::runtime_error("Incorrect level statistics");
        }

        totalHitCount += levelStatistics.amountHits;
        expectedLevelRequests = levelStatistics.amountMisses;
    }

    if (expectedLevelRequests != memoryLoadCount ||
        totalHitCount + memoryLoadCount != requestCount ||
        systemStatistics.total.amountRequests != requestCount ||
        systemStatistics.total.amountHits != totalHitCount ||
        systemStatistics.total.amountMisses != memoryLoadCount) {
        throw std::runtime_error("Incorrect total statistics");
    }
}

cache::CacheSystemStats runCacheExperiment(const cache::cacheSystemParams& configuration,
                                          const std::vector<int>& requestSequence) {

    if (requestSequence.empty()) {
        throw std::invalid_argument("Empty request list");
    }

    cache::CacheSystem<uint32_t, int> cacheSystem(configuration);
    std::size_t memoryLoadCount = 0;

    auto slowGetPage = [&](int requestedKey) {
        ++memoryLoadCount;
        return generatePageValue(requestedKey);
    };

    for (const int requestedKey : requestSequence) {
        if (cacheSystem.lookupUpdate(requestedKey, slowGetPage) != generatePageValue(requestedKey)) {
            throw std::runtime_error("Incorrect page");
        }
    }

    const auto systemStatistics = cacheSystem.getStats();

    checkSystemStatistics(
        systemStatistics, requestSequence.size(), memoryLoadCount, configuration.levels.size());

    return systemStatistics;
}

double calculateAverageAccessCost(const cache::CacheSystemStats& systemStatistics) {
    double totalAccessCost = systemStatistics.total.amountMisses * kMemoryAccessCost;

    for (std::size_t levelIndex = 0; levelIndex < systemStatistics.levels.size(); ++levelIndex) {
        totalAccessCost += systemStatistics.levels[levelIndex].stats.amountRequests *
                           kLevelAccessCosts[levelIndex];
    }

    return totalAccessCost / systemStatistics.total.amountRequests;
}

double calculateHitRate(const cache::CacheSystemStats& systemStatistics) {
    return static_cast<double>(systemStatistics.total.amountHits) /
           systemStatistics.total.amountRequests;
}


void writeCsvHeader(std::ostream& csvOutput) {
    csvOutput << std::setprecision(12)
              << "pattern,seed,configuration,requests,"
                 "l1_requests,l1_hits,l1_misses,l2_requests,l2_hits,l2_misses,"
                 "l3_requests,l3_hits,l3_misses,memory_misses,hit_rate,amat\n";
}

void writeCsvResult(std::ostream& csvOutput,
                    const std::string& patternName,
                    unsigned randomSeed,
                    const cache::cacheSystemParams& configuration,
                    const cache::CacheSystemStats& systemStatistics) {

    csvOutput << patternName << ',' << randomSeed << ','
              << getConfigurationName(configuration) << ','
              << systemStatistics.total.amountRequests;

    for (std::size_t levelIndex = 0; levelIndex < kCacheLevels.size(); ++levelIndex) {
        if (levelIndex < systemStatistics.levels.size()) {
            const auto& levelStatistics = systemStatistics.levels[levelIndex].stats;

            csvOutput << ',' << levelStatistics.amountRequests
                      << ',' << levelStatistics.amountHits
                      << ',' << levelStatistics.amountMisses;
        } else {
            csvOutput << ",0,0,0";
        }
    }

    csvOutput << ',' << systemStatistics.total.amountMisses
              << ',' << calculateHitRate(systemStatistics)
              << ',' << calculateAverageAccessCost(systemStatistics) << '\n';
}


std::vector<std::size_t> rankConfigurationScores(const std::vector<double>& configurationScores) {
    std::vector<std::size_t> rankedIndices;
    std::vector<bool> selectedConfigurations(configurationScores.size(), false);

    for (std::size_t rankIndex = 0; rankIndex < configurationScores.size(); ++rankIndex) {
        std::size_t bestConfigurationIndex = configurationScores.size();

        for (std::size_t configurationIndex = 0;
             configurationIndex < configurationScores.size();
             ++configurationIndex) {

            if (!selectedConfigurations[configurationIndex] &&
                (bestConfigurationIndex == configurationScores.size() ||
                 configurationScores[configurationIndex] < configurationScores[bestConfigurationIndex])) {
                bestConfigurationIndex = configurationIndex;
            }
        }

        selectedConfigurations[bestConfigurationIndex] = true;
        rankedIndices.push_back(bestConfigurationIndex);
    }

    return rankedIndices;
}

void printBestConfigurations(const std::vector<cache::cacheSystemParams>& cacheConfigurations,
                             const std::vector<double>& configurationScores,
                             const std::string& metricName,
                             std::size_t resultCount,
                             const std::vector<double>& meanHitRates) {

    const auto rankedIndices = rankConfigurationScores(configurationScores);

    for (std::size_t rankIndex = 0;
         rankIndex < resultCount && rankIndex < rankedIndices.size();
         ++rankIndex) {

        const auto configurationIndex = rankedIndices[rankIndex];

        std::cout << "  " << rankIndex + 1 << ". "
                  << getConfigurationName(cacheConfigurations[configurationIndex])
                  << ' ' << metricName << '=' << configurationScores[configurationIndex];

        if (!meanHitRates.empty()) {
            std::cout << " hit=" << 100 * meanHitRates[configurationIndex] << '%';
        }

        std::cout << '\n';
    }

    for (const auto configurationIndex : rankedIndices) {
        if (cacheConfigurations[configurationIndex].levels.size() > 1) {
            std::cout << "  Best multi-level: "
                      << getConfigurationName(cacheConfigurations[configurationIndex])
                      << ' ' << metricName << '=' << configurationScores[configurationIndex] << '\n';
            break;
        }
    }
}


std::vector<double> evaluateRequestPattern(
    const RequestPatternDescription& requestPattern,
    const std::vector<cache::cacheSystemParams>& cacheConfigurations,
    std::ostream& csvOutput) {

    std::cout << '\n' << requestPattern.patternName << " ...\n" << std::flush;

    std::vector<double> meanAccessCosts(cacheConfigurations.size(), 0.0);
    std::vector<double> meanHitRates(cacheConfigurations.size(), 0.0);

    for (const auto randomSeed : kRandomSeeds) {
        const auto requestSequence = generateRequestSequence(requestPattern.patternType, randomSeed);

        for (std::size_t configurationIndex = 0;
             configurationIndex < cacheConfigurations.size();
             ++configurationIndex) {

            try {
                const auto systemStatistics = runCacheExperiment(
                    cacheConfigurations[configurationIndex], requestSequence);

                meanAccessCosts[configurationIndex] +=
                    calculateAverageAccessCost(systemStatistics) / kRandomSeeds.size();

                meanHitRates[configurationIndex] +=
                    calculateHitRate(systemStatistics) / kRandomSeeds.size();

                writeCsvResult(csvOutput, requestPattern.patternName, randomSeed,
                               cacheConfigurations[configurationIndex], systemStatistics);
            } catch (const std::exception& experimentError) {
                throw std::runtime_error(
                    requestPattern.patternName + ", seed=" + std::to_string(randomSeed) +
                    ", " + getConfigurationName(cacheConfigurations[configurationIndex]) +
                    ": " + experimentError.what());
            }
        }
    }

    printBestConfigurations(cacheConfigurations, meanAccessCosts, "AMAT", 3, meanHitRates);

    return meanAccessCosts;
}

void updateAverageSlowdown(const std::vector<double>& meanAccessCosts,
                           std::vector<double>& meanSlowdownPercent) {

    double minimumPatternCost = meanAccessCosts[0];

    for (const double meanAccessCost : meanAccessCosts) {
        if (meanAccessCost < minimumPatternCost) {
            minimumPatternCost = meanAccessCost;
        }
    }

    for (std::size_t configurationIndex = 0;
         configurationIndex < meanAccessCosts.size();
         ++configurationIndex) {

        meanSlowdownPercent[configurationIndex] +=
            100 * (meanAccessCosts[configurationIndex] / minimumPatternCost - 1) /
            kRequestPatterns.size();
    }
}

void runCacheBenchmark(const std::string& outputFilePath) {
    std::ofstream csvOutput(outputFilePath);

    if (!csvOutput) {
        throw std::runtime_error("Cannot open CSV: " + outputFilePath);
    }

    writeCsvHeader(csvOutput);

    const auto cacheConfigurations = createCacheConfigurations();
    std::vector<double> meanSlowdownPercent(cacheConfigurations.size(), 0.0);

    std::cout << std::fixed << std::setprecision(3)
              << cacheConfigurations.size() << " configurations, "
              << kRequestsPerTrace << " requests, "
              << kRandomSeeds.size() << " seeds; cold start included.\n";

    for (const auto& requestPattern : kRequestPatterns) {
        const auto meanAccessCosts = evaluateRequestPattern(
            requestPattern, cacheConfigurations, csvOutput);

        updateAverageSlowdown(meanAccessCosts, meanSlowdownPercent);
    }

    std::cout << "\nTop 5 compromises: mean slowdown (%) vs each pattern's best\n";
    printBestConfigurations(cacheConfigurations, meanSlowdownPercent, "slowdown(%)", 5, {});

    csvOutput.close();

    if (!csvOutput) {
        throw std::runtime_error("Failed to write CSV: " + outputFilePath);
    }

    std::cout << "\nCSV: " << outputFilePath
              << "\nBest among these candidates under this cost model.\n";
}

}


int main(int argc, char** argv) {
    try {
        if (argc > 2) {
            throw std::invalid_argument("Usage: cache_benchmark [output.csv]");
        }

        std::string outputFilePath = "cache_benchmark_results.csv";

        if (argc == 2) {
            outputFilePath = argv[1];
        }

        benchmark::runCacheBenchmark(outputFilePath);
    } catch (const std::exception& benchmarkError) {
        std::cerr << "Benchmark failed: " << benchmarkError.what() << '\n';
        return 1;
    }

    return 0;
}

File: cacheBenchmarks/cache_benchmark.py
import argparse
import csv
import itertools
import random
import subprocess
from pathlib import Path


CACHE_STRATEGIES = ("LRU", "LFU", "ARC", "2Q", "LIRS")
LEVEL_CAPACITIES = ((64,), (8, 56), (16, 48), (32, 32), (8, 16, 40))
LEVEL_ACCESS_COSTS = (1.0, 5.0, 20.0)
MEMORY_ACCESS_COST = 100.0

REQUEST_PATTERNS = (
    "small_cycle", "boundary_cycle", "scan", "hot_cold",
    "hot_scan", "phase_change", "bursts", "uniform",
)

CSV_COLUMNS = (
    "pattern", "seed", "configuration", "requests",
    "l1_requests", "l1_hits", "l1_misses",
    "l2_requests", "l2_hits", "l2_misses",
    "l3_requests", "l3_hits", "l3_misses",
    "memory_misses", "hit_rate", "amat",
    "ideal_memory_misses", "ideal_hit_rate", "ideal_amat",
    "extra_memory_misses", "slowdown_vs_ideal_pct",
)


def generate_hot_cold_key(random_generator, hot_page_offset):
    if random_generator.randrange(100) < 90:
        return hot_page_offset + random_generator.randrange(8)

    return random_generator.randint(1024, 2047)


def generate_request_key(pattern_name, request_index, previous_key, random_generator):
    if pattern_name == "small_cycle":
        return request_index % 32
    if pattern_name == "boundary_cycle":
        return request_index % 65
    if pattern_name == "scan":
        return request_index % 256
    if pattern_name == "hot_cold":
        return generate_hot_cold_key(random_generator, 0)
    if pattern_name == "hot_scan":
        if request_index % 512 < 384:
            return random_generator.randrange(8)
        return 1024 + request_index
    if pattern_name == "phase_change":
        return generate_hot_cold_key(random_generator, (request_index // 2000) % 4 * 64)
    if pattern_name == "bursts":
        if request_index % 16 == 0:
            return random_generator.randrange(256)
        return previous_key
    if pattern_name == "uniform":
        return random_generator.randrange(256)

    raise ValueError(f"Unknown pattern: {pattern_name}")


def generate_request_sequence(pattern_name, random_seed, request_count):
    random_generator = random.Random(random_seed)
    request_sequence = []
    previous_key = 0

    for request_index in range(request_count):
        previous_key = generate_request_key(
            pattern_name, request_index, previous_key, random_generator
        )
        request_sequence.append(previous_key)

    return request_sequence


def create_cache_configurations():
    configurations = []

    for capacities in LEVEL_CAPACITIES:
        for strategies in itertools.product(CACHE_STRATEGIES, repeat=len(capacities)):
            strategies = tuple(reversed(strategies))
            configuration_name = "+".join(
                strategy + str(capacity)
                for strategy, capacity in zip(strategies, capacities)
            )
            configurations.append((configuration_name, strategies, capacities))

    return configurations


def get_config_path(root_directory, configuration_name):
    return root_directory / "configs" / f"benchmark_{configuration_name}.txt"


def get_input_path(root_directory, pattern_name, random_seed, capacities):
    capacity_name = "_".join(str(capacity) for capacity in capacities)
    return root_directory / "inputs" / (
        f"benchmark_{pattern_name}_seed{random_seed}_{capacity_name}.txt"
    )


def write_config_file(file_path, strategies):
    file_path.write_text(
        str(len(strategies)) + "\n" + "\n".join(strategies) + "\n",
        encoding="utf-8",
    )


def write_input_file(file_path, capacities, requests):
    capacity_line = " ".join(str(capacity) for capacity in capacities)
    request_line = " ".join(str(request) for request in requests)
    file_path.write_text(
        f"{capacity_line}\n{len(requests)}\n{request_line}\n", encoding="utf-8"
    )


def generate_benchmark_files(root_directory, configurations, random_seeds, request_count):
    (root_directory / "configs").mkdir(parents=True, exist_ok=True)
    (root_directory / "inputs").mkdir(parents=True, exist_ok=True)

    for configuration_name, strategies, capacities in configurations:
        write_config_file(get_config_path(root_directory, configuration_name), strategies)

    write_config_file(get_config_path(root_directory, "REF64"), ("REF",))

    for pattern_name in REQUEST_PATTERNS:
        for random_seed in random_seeds:
            requests = generate_request_sequence(pattern_name, random_seed, request_count)

            for capacities in LEVEL_CAPACITIES:
                input_path = get_input_path(root_directory, pattern_name, random_seed, capacities)
                write_input_file(input_path, capacities, requests)


def check_statistics(statistics, request_count, level_count):
    if len(statistics) != 12 or any(value < 0 for value in statistics):
        raise ValueError("Runner must return 12 nonnegative integer counters")

    total_requests, total_hits, total_misses = statistics[:3]
    expected_requests = request_count
    accumulated_hits = 0

    for level_index in range(3):
        offset = 3 + 3 * level_index
        level_requests, level_hits, level_misses = statistics[offset:offset + 3]

        if level_index >= level_count:
            if level_requests or level_hits or level_misses:
                raise ValueError("Nonzero statistics for an absent level")
            continue

        if level_requests != expected_requests or level_hits + level_misses != level_requests:
            raise ValueError("Incorrect level statistics")

        accumulated_hits += level_hits
        expected_requests = level_misses

    if (total_requests != request_count or total_hits != accumulated_hits
            or total_misses != expected_requests or total_hits + total_misses != total_requests):
        raise ValueError("Incorrect total statistics")


def run_cache_experiment(runner_path, config_path, input_path, request_count, level_count):
    try:
        process_result = subprocess.run(
            [str(runner_path), str(config_path), str(input_path)],
            capture_output=True, text=True, check=True,
        )
        statistics = [int(value) for value in process_result.stdout.split()]
        check_statistics(statistics, request_count, level_count)
        return statistics
    except subprocess.CalledProcessError as error:
        raise RuntimeError(
            f"{config_path.name}, {input_path.name}: "
            f"{error.stderr.strip() or error.stdout.strip() or error}"
        ) from error
    except ValueError as error:
        raise RuntimeError(f"{config_path.name}, {input_path.name}: {error}") from error


def calculate_average_access_cost(statistics):
    total_cost = statistics[2] * MEMORY_ACCESS_COST

    for level_index, access_cost in enumerate(LEVEL_ACCESS_COSTS):
        total_cost += statistics[3 + 3 * level_index] * access_cost

    return total_cost / statistics[0]


def calculate_hit_rate(statistics):
    return statistics[1] / statistics[0]


def calculate_slowdown_percent(access_cost, reference_cost):
    return 100 * (access_cost / reference_cost - 1)


def write_csv_result(csv_writer, pattern_name, random_seed, configuration_name,
                     statistics, ideal_statistics):
    access_cost = calculate_average_access_cost(statistics)
    ideal_cost = calculate_average_access_cost(ideal_statistics)

    csv_writer.writerow([
        pattern_name, random_seed, configuration_name, statistics[0],
        *statistics[3:], statistics[2], calculate_hit_rate(statistics), access_cost,
        ideal_statistics[2], calculate_hit_rate(ideal_statistics), ideal_cost,
        statistics[2] - ideal_statistics[2],
        calculate_slowdown_percent(access_cost, ideal_cost),
    ])


def format_configuration_result(configuration_name, score, metric_name,
                                 hit_rate=None, ideal_cost=None):
    result = f"{configuration_name} {metric_name}={score:.3f}"

    if hit_rate is not None:
        result += f" hit={100 * hit_rate:.3f}%"
    if ideal_cost is not None:
        result += f" vs REF={calculate_slowdown_percent(score, ideal_cost):.3f}%"

    return result


def print_best_configurations(configurations, scores, metric_name, result_count,
                              hit_rates=None, ideal_cost=None):
    ranked_indices = sorted(range(len(configurations)), key=lambda index: scores[index])

    for rank, configuration_index in enumerate(ranked_indices[:result_count], start=1):
        hit_rate = None if hit_rates is None else hit_rates[configuration_index]
        result = format_configuration_result(
            configurations[configuration_index][0], scores[configuration_index],
            metric_name, hit_rate, ideal_cost,
        )
        print(f"  {rank}. {result}")

    for configuration_index in ranked_indices:
        if len(configurations[configuration_index][2]) > 1:
            result = format_configuration_result(
                configurations[configuration_index][0], scores[configuration_index],
                metric_name, ideal_cost=ideal_cost,
            )
            print(f"  Best multi-level: {result}")
            break


def evaluate_request_pattern(pattern_name, configurations, arguments, csv_writer):
    print(f"\n{pattern_name} ...", flush=True)
    mean_costs = [0.0] * len(configurations)
    mean_hit_rates = [0.0] * len(configurations)
    mean_ideal_cost = 0.0
    mean_ideal_hit_rate = 0.0
    mean_ideal_misses = 0.0
    seed_count = len(arguments.seeds)

    for random_seed in arguments.seeds:
        ideal_statistics = run_cache_experiment(
            arguments.runner, get_config_path(arguments.root, "REF64"),
            get_input_path(arguments.root, pattern_name, random_seed, (64,)),
            arguments.requests, 1,
        )
        mean_ideal_cost += calculate_average_access_cost(ideal_statistics) / seed_count
        mean_ideal_hit_rate += calculate_hit_rate(ideal_statistics) / seed_count
        mean_ideal_misses += ideal_statistics[2] / seed_count
        write_csv_result(csv_writer, pattern_name, random_seed, "REF64",
                         ideal_statistics, ideal_statistics)

        for configuration_index, configuration in enumerate(configurations):
            configuration_name, strategies, capacities = configuration
            statistics = run_cache_experiment(
                arguments.runner, get_config_path(arguments.root, configuration_name),
                get_input_path(arguments.root, pattern_name, random_seed, capacities),
                arguments.requests, len(capacities),
            )
            if statistics[2] < ideal_statistics[2]:
                raise RuntimeError(
                    f"{pattern_name}, seed={random_seed}, {configuration_name}: "
                    "fewer misses than REF64; check cache implementations"
                )

            mean_costs[configuration_index] += calculate_average_access_cost(statistics) / seed_count
            mean_hit_rates[configuration_index] += calculate_hit_rate(statistics) / seed_count
            write_csv_result(csv_writer, pattern_name, random_seed, configuration_name,
                             statistics, ideal_statistics)

    print(f"  Ideal REF64 AMAT={mean_ideal_cost:.3f} "
          f"hit={100 * mean_ideal_hit_rate:.3f}% mean misses={mean_ideal_misses:.3f}")
    print_best_configurations(configurations, mean_costs, "AMAT", 3,
                              mean_hit_rates, mean_ideal_cost)
    return mean_costs


def run_cache_benchmark(configurations, arguments):
    mean_slowdown = [0.0] * len(configurations)
    print(f"{len(configurations)} configurations, {arguments.requests} requests, "
          f"{len(arguments.seeds)} seeds; cold start included.")
    print("Reference: offline REF64, one level; same requests and total capacity.")

    with arguments.output.open("w", newline="", encoding="utf-8") as output_file:
        csv_writer = csv.writer(output_file)
        csv_writer.writerow(CSV_COLUMNS)

        for pattern_name in REQUEST_PATTERNS:
            mean_costs = evaluate_request_pattern(pattern_name, configurations, arguments, csv_writer)
            best_pattern_cost = min(mean_costs)

            for configuration_index, mean_cost in enumerate(mean_costs):
                mean_slowdown[configuration_index] += (
                    calculate_slowdown_percent(mean_cost, best_pattern_cost) / len(REQUEST_PATTERNS)
                )

    print("\nTop 5 compromises: mean slowdown (%) vs each pattern's best")
    print_best_configurations(configurations, mean_slowdown, "slowdown(%)", 5)
    print(f"\nCSV: {arguments.output}")
    print("Best among these candidates under this cost model.")


def read_arguments():
    argument_parser = argparse.ArgumentParser(description="Generate cache traces and run benchmarks")
    argument_parser.add_argument("--runner", type=Path, help="Path to cache_benchmark_runner")
    argument_parser.add_argument("--root", type=Path, default=Path("./cacheBenchmarks"), help="Directory for configs/inputs")
    argument_parser.add_argument("--requests", type=int, default=20000)
    argument_parser.add_argument("--seeds", type=int, nargs="+", default=[1, 2, 3])
    argument_parser.add_argument("--output", type=Path, default=Path("./cacheBenchmarks/cache_benchmark_results.csv"))
    argument_parser.add_argument("--generate-only", action="store_true")
    arguments = argument_parser.parse_args()

    if not 1 <= arguments.requests <= 2147482623:
        argument_parser.error("--requests must be positive and fit the parser's integer range")
    if len(arguments.seeds) != len(set(arguments.seeds)):
        argument_parser.error("--seeds must contain distinct values")

    arguments.root = arguments.root.resolve()
    if not arguments.generate_only:
        if arguments.runner is None:
            argument_parser.error("Pass --runner or use --generate-only")
        arguments.runner = arguments.runner.resolve()
        if not arguments.runner.is_file():
            argument_parser.error(f"Runner not found: {arguments.runner}")

    return arguments


def main():
    arguments = read_arguments()
    configurations = create_cache_configurations()
    generate_benchmark_files(arguments.root, configurations, arguments.seeds, arguments.requests)

    if arguments.generate_only:
        input_count = len(REQUEST_PATTERNS) * len(arguments.seeds) * len(LEVEL_CAPACITIES)
        print(f"Generated {len(configurations) + 1} configs and {input_count} inputs in {arguments.root}")
    else:
        run_cache_benchmark(configurations, arguments)


if __name__ == "__main__":
    try:
        main()
    except (OSError, RuntimeError, ValueError) as benchmark_error:
        raise SystemExit(f"Benchmark failed: {benchmark_error}")
